plugins/inputs/slurm/slurm.go

```diff
@@ -0,0 +1,473 @@
+//go:generate ../../../tools/config_includer/generator
+//go:generate ../../../tools/readme_config_includer/generator
+package slurm
+
+import (
+	"context"
+	_ "embed"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+
+	goslurm "github.com/pcolladosoto/goslurm/v0038"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type Slurm struct {
+	URL              string          `toml:"url"`
+	Username         string          `toml:"username"`
+	Token            string          `toml:"token"`
+	EnabledEndpoints []string        `toml:"enabled_endpoints"`
+	ResponseTimeout  config.Duration `toml:"response_timeout"`
+	Log              telegraf.Logger `toml:"-"`
+	tls.ClientConfig
+
+	client      *goslurm.APIClient
+	baseURL     *url.URL
+	endpointMap map[string]bool
+}
+
+func (*Slurm) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *Slurm) Init() error {
+	if s.EnabledEndpoints == nil {
+		s.EnabledEndpoints = []string{"diag", "jobs", "nodes", "partitions", "reservations"}
+	}
+
+	s.endpointMap = map[string]bool{}
+	for _, endpoint := range s.EnabledEndpoints {
+		s.endpointMap[strings.ToLower(endpoint)] = true
+	}
+
+	if s.URL == "" {
+		return errors.New("empty URL provided")
+	}
+
+	u, err := url.Parse(s.URL)
+	if err != nil {
+		return err
+	}
+
+	if u.Hostname() == "" {
+		return fmt.Errorf("empty hostname for url %q", s.URL)
+	}
+
+	s.baseURL = u
+
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return fmt.Errorf("invalid scheme %q", u.Scheme)
+	}
+
+	tlsCfg, err := s.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	if u.Scheme == "http" && tlsCfg != nil {
+		s.Log.Warn("non-empty TLS configuration for a URL with an http scheme. Ignoring it...")
+		tlsCfg = nil
+	}
+
+	configuration := goslurm.NewConfiguration()
+	configuration.Host = u.Host
+	configuration.Scheme = u.Scheme
+	configuration.UserAgent = internal.ProductToken()
+	configuration.HTTPClient = &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+		},
+		Timeout: time.Duration(s.ResponseTimeout),
+	}
+
+	s.client = goslurm.NewAPIClient(configuration)
+
+	return nil
+}
+
+func (s *Slurm) parseTres(tres string) map[string]interface{} {
+	parsedValues := map[string]interface{}{}
+
+	for _, tresVal := range strings.Split(tres, ",") {
+		parsedTresVal := strings.Split(tresVal, "=")
+		if len(parsedTresVal) != 2 {
+			continue
+		}
+
+		tag := parsedTresVal[0]
+		val := parsedTresVal[1]
+		var (
+			factor float64 = 1
+			ok     bool
+		)
+
+		if tag == "mem" {
+			factor, ok = map[string]float64{
+				"K": 1.0 / 1024.0,
+				"M": 1,
+				"G": 1024,
+				"T": 1024 * 1024,
+			}[strings.ToUpper(val[len(val)-1:])]
```

Comment:
As seen [here](https://slurm.schedmd.com/archive/slurm-22.05.9/tres.html) (on the **TRESBillingWeights** section) the base unit for memory is megabytes. What's more, only the K, M, G, T and P suffixes are mentioned.

I went ahead and added support for the P suffix just in case. Regarding the use of bytes I'm not really sure that's something that we might encounter on a real system... What do you think?

Reply:
Let's keep it this way then. I only considered the corner case of `mem=0` which might cause a panic... But given the documentation does not mention this I guess it's ok as-is.